lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -91,6 +94,45 @@ sub view ($self) {
     $self->viewimg;
 }
 
+sub _checkout_needledir_or_default {
+    my $job = shift;
+    my $tmp_needles_dir = $job->needle_dir;
+    my $real_needle_dir = realpath($tmp_needles_dir) // $tmp_needles_dir;
+    if (my $casedir = $job->settings->single({key => 'CASEDIR'})) {
+        my $casedir_refspec = Mojo::URL->new($casedir->value)->fragment;
+        eval {
+            my $vars_json = Mojo::File->new($job->result_dir(), 'vars.json')->slurp;
+            my $vars = decode_json($vars_json);
+            $casedir_refspec = $vars->{TEST_GIT_HASH};
+        };
+        qx{git -C "$real_needle_dir" pull};
+        my $current_repo_sha = qx{git -C "$real_needle_dir" rev-parse HEAD};
+        chomp($current_repo_sha);
+        chomp($casedir_refspec);
+        if ($casedir_refspec eq "" || $current_repo_sha eq $casedir_refspec) {
+            $tmp_needles_dir = $real_needle_dir;
+        }
+        else {
+            my $needle_dir_basename = basename($real_needle_dir);
+            my $new_path = "/tmp/$needle_dir_basename/worktrees/$casedir_refspec";
+            if (!-d $new_path) {
+                my $git_wt_cmd = "git -C $real_needle_dir worktree";
+                # See how many worktrees we have already checked out, if there are too many, remove them all
+                make_path("/tmp/$needle_dir_basename/worktrees") unless (-d "/tmp/$needle_dir_basename/worktrees");
+                opendir my $dh, "/tmp/$needle_dir_basename/worktrees";
+                my $num_entries = () = readdir($dh);
+                if ($num_entries > 8) {
+                    qx{$git_wt_cmd list | grep /tmp | awk '{ print \$1 }' | xargs -n1 $git_wt_cmd remove &>/dev/null};
+                }
```

Comment:
That is not going to cut it in our production use. The number 8 needed to be configurable and even then we must really avoid removing worktrees that are still in use. It would also make more sense to do the cleanup as a background task (we use Minion jobs for this) instead in the context of serving an HTTP response (where we can still *trigger* the cleanup).